File: elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer
  def score(category, dice) do
    case category do
      :choice          -> Enum.sum(dice)
      :ones            -> Enum.count(dice, &(&1 == 1)) * 1
      :twos            -> Enum.count(dice, &(&1 == 2)) * 2
      :threes          -> Enum.count(dice, &(&1 == 3)) * 3
      :fours           -> Enum.count(dice, &(&1 == 4)) * 4
      :fives           -> Enum.count(dice, &(&1 == 5)) * 5
      :sixes           -> Enum.count(dice, &(&1 == 6)) * 6
      :yacht           -> if dice |> Enum.uniq |> Enum.count == 1, do: 50, else: 0
      :little_straight -> if Enum.sort(dice) == [1, 2, 3, 4, 5], do: 30, else: 0
      :big_straight    -> if Enum.sort(dice) == [2, 3, 4, 5, 6], do: 30, else: 0
      :full_house      -> if Enum.frequencies(dice) |> Map.values |> Enum.sort == [2, 3], do: Enum.sum(dice), else: 0
      :four_of_a_kind  -> Enum.frequencies(dice) |> Enum.map(fn {k, v} -> if v in [4, 5], do: k * 4, else: 0 end) |> Enum.sum
    end
  end
end

File: csharp/yacht/Yacht.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum YachtCategory
{
    Ones           = 1,
    Twos           = 2,
    Threes         = 3,
    Fours          = 4,
    Fives          = 5,
    Sixes          = 6,
    FullHouse      = 7,
    FourOfAKind    = 8,
    LittleStraight = 9,
    BigStraight    = 10,
    Choice         = 11,
    Yacht          = 12,
}

public static class YachtGame
{
    public static int Score(int[] dice, YachtCategory category)
    {
        if ((int)category <= 6) return dice.Where(x => x == (int)category).Count() * (int)category;
        if ((int)category == 7 && IsFullHouse(dice)) return dice.Sum();
        if ((int)category == 8) return Frequencies(dice).Select(kvp => kvp.Value == 4 || kvp.Value == 5 ? 4 * kvp.Key : 0).Sum();
        if ((int)category == 9 && IsStraight(dice) && dice.Contains(1)) return 30;
        if ((int)category == 10 && IsStraight(dice) && dice.Contains(6)) return 30;
        if ((int)category == 11) return dice.Sum();
        if ((int)category == 12 && dice.Distinct().Count() == 1) return 50;
        return 0;
    }

    static bool IsFullHouse(int[] dice)
    {
        List<int> fhFreqs = new List<int> { 2, 3 };
        return Frequencies(dice).Select(x => x.Value).OrderBy(x => x).ToList().SequenceEqual(fhFreqs);
    }

    static bool IsStraight(int[] dice)
    {
        return dice.Distinct().Count() == 5 && (!dice.Contains(1) || !dice.Contains(6));
    }

    static Dictionary<int, int> Frequencies(int[] dice)
    {
        return dice.GroupBy(x => x).ToDictionary(x => x.Key, x => x.Count());
    }
}

